File: kid-ios/MyWorld/Parent/SchedulesView.swift
import SwiftUI

/// PRD §4.6 — the scheduling manager, v1 scope: see every prompt the child's
/// board will fire, flip them on/off, delete, and add a simple reminder.
/// Question-with-responses and game-sequence authoring stay on the web (they
/// need the tile pickers); this screen round-trips those rows untouched.
///
/// Works on raw dictionaries rather than the typed Schedule model so fields
/// this version doesn't know about (responses, scopes, …) survive the write.
struct SchedulesView: View {
    @Environment(AuthManager.self) private var auth

    @State private var rows: [[String: Any]] = []
    @State private var loaded = false
    @State private var saving = false
    @State private var showNew = false

    private let api = APIClient()

    var body: some View {
        List {
            if loaded && rows.isEmpty {
                Text("No scheduled prompts yet. Add a reminder below, or build richer prompts (questions, game nudges) on the web dashboard.")
                    .font(.footnote).foregroundStyle(.secondary)
            }
            ForEach(Array(rows.enumerated()), id: \.offset) { idx, row in
                scheduleRow(idx: idx, row: row)
            }
            .onDelete { offsets in
                rows.remove(atOffsets: offsets)
                Task { await save() }
            }
        }
        .navigationTitle("Schedules")
        .navigationBarTitleDisplayMode(.inline)
        .toolbar {
            ToolbarItem(placement: .topBarTrailing) {
                Button { showNew = true } label: { Image(systemName: "plus") }
            }
        }
        .sheet(isPresented: $showNew) {
            NewReminderSheet { prompt, intervalMin in
                rows.append([
                    "id": "s\(Int(Date().timeIntervalSince1970 * 1000))",
                    "type": "reminder",
                    "enabled": true,
                    "prompt": prompt,
                    "timing": "interval",
                    "intervalMin": intervalMin,
                    "days": [0, 1, 2, 3, 4, 5, 6],
                ])
                Task { await save() }
            }
        }
        .overlay(alignment: .bottom) {
            if saving {
                Text("Saving…").font(.footnote).foregroundStyle(.secondary)
                    .padding(8).background(.thinMaterial, in: Capsule()).padding(.bottom, 8)
            }
        }
        .task {
            rows = await api.fetchRawSchedules(childId: auth.childSlug)
            loaded = true
        }
    }

    private func scheduleRow(idx: Int, row: [String: Any]) -> some View {
        let type = (row["type"] as? String) ?? "reminder"
        let prompt = (row["prompt"] as? String).flatMap { $0.isEmpty ? nil : $0 } ?? defaultPrompt(type)
        let enabled = (row["enabled"] as? Bool) ?? true
        let when = timingText(row)
        return HStack {
            Image(systemName: icon(type))
                .foregroundStyle(Color(hex: "#ff1493"))
                .frame(width: 28)
            VStack(alignment: .leading, spacing: 2) {
                Text(prompt).font(.system(size: 15, weight: .semibold))
                Text("\(label(type)) · \(when)").font(.system(size: 12)).foregroundStyle(.secondary)
            }
            Spacer()
            Toggle("", isOn: Binding(
                get: { (rows[idx]["enabled"] as? Bool) ?? true },
                set: { newVal in
                    rows[idx]["enabled"] = newVal
                    Task { await save() }
                }
            ))
            .labelsHidden()
            .tint(Color(hex: "#ff1493"))
        }
        .opacity(enabled ? 1 : 0.55)
    }

    private func icon(_ type: String) -> String {
        switch type {
        case "question": return "questionmark.bubble.fill"
        case "game":     return "gamecontroller.fill"
        default:           return "bell.fill"
        }
    }
    private func label(_ type: String) -> String {
        switch type {
        case "question": return "Question"
        case "game":     return "Game nudge"
        default:           return "Reminder"
        }
    }
    private func defaultPrompt(_ type: String) -> String {
        switch type {
        case "question": return "A quick question."
        case "game":     return "Let's do a game!"
        default:           return "Time for a check-in."
        }
    }
    private func timingText(_ row: [String: Any]) -> String {
        if (row["timing"] as? String) == "times", let times = row["times"] as? [String], !times.isEmpty {
            return "at " + times.joined(separator: ", ")
        }
        let mins = (row["intervalMin"] as? Double) ?? Double(row["intervalMin"] as? Int ?? 45)
        return "every \(Int(mins)) min"
    }

    private func save() async {
        saving = true
        await api.saveSchedules(childId: auth.childSlug, rows)
        saving = false
    }
}

/// Minimal composer for a spoken reminder ("Do you need the potty?") on an
/// interval. The board speaks it in the parent's recorded reward voice.
private struct NewReminderSheet: View {
    let onAdd: (String, Int) -> Void
    @Environment(\.dismiss) private var dismiss
    @State private var prompt = ""
    @State private var intervalMin = 45

    var body: some View {
        NavigationStack {
            Form {
                TextField("What should the board say?", text: $prompt, axis: .vertical)
                    .lineLimit(1...3)
                Stepper("Every \(intervalMin) minutes", value: $intervalMin, in: 5...240, step: 5)
            }
            .navigationTitle("New reminder")
            .navigationBarTitleDisplayMode(.inline)
            .toolbar {
                ToolbarItem(placement: .topBarLeading) { Button("Cancel") { dismiss() } }
                ToolbarItem(placement: .topBarTrailing) {
                    Button("Add") {
                        onAdd(prompt.trimmingCharacters(in: .whitespaces), intervalMin)
                        dismiss()
                    }
                    .disabled(prompt.trimmingCharacters(in: .whitespaces).isEmpty)
                }
            }
        }
        .presentationDetents([.medium])
    }
}

File: android-native/app/src/main/java/io/andrewpeterson/myworld/ui/parent/SchedulesView.kt
package io.andrewpeterson.myworld.ui.parent

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.AlertDialog
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Switch
import androidx.compose.material3.SwitchDefaults
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.compose.ui.window.Dialog
import androidx.compose.ui.window.DialogProperties
import io.andrewpeterson.myworld.LocalAppContainer
import io.andrewpeterson.myworld.net.childSettings
import io.andrewpeterson.myworld.net.saveChildSettingsKey
import io.andrewpeterson.myworld.ui.theme.Brand
import io.andrewpeterson.myworld.ui.theme.hexColor
import kotlinx.coroutines.launch
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.boolean
import kotlinx.serialization.json.buildJsonArray
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.doubleOrNull
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.put

/**
 * Scheduling manager — port of `Parent/SchedulesView.swift` (PRD §4.6 v1):
 * list every prompt the board will fire, flip on/off, delete, add a simple
 * interval reminder. Question/game authoring stays on the web; those rows
 * round-trip UNTOUCHED because everything here works on raw JsonObjects.
 */
@Composable
fun SchedulesView(onDismiss: () -> Unit) {
    val c = LocalAppContainer.current
    val scope = rememberCoroutineScope()

    var rows by remember { mutableStateOf<List<JsonObject>?>(null) }
    var saving by remember { mutableStateOf(false) }
    var showNew by remember { mutableStateOf(false) }

    LaunchedEffect(Unit) {
        rows = try {
            (c.api.childSettings(c.auth.childSlug)["schedules"] as? JsonArray)
                ?.mapNotNull { it as? JsonObject } ?: emptyList()
        } catch (_: Exception) { emptyList() }
    }

    fun save(newRows: List<JsonObject>) {
        rows = newRows
        saving = true
        scope.launch {
            c.api.saveChildSettingsKey(c.auth.childSlug, "schedules",
                buildJsonArray { newRows.forEach { add(it) } })
            saving = false
        }
    }

    Dialog(onDismissRequest = onDismiss, properties = DialogProperties(usePlatformDefaultWidth = false)) {
        Column(Modifier.fillMaxSize().background(hexColor("#fff7fb"))) {
            Row(
                Modifier.fillMaxWidth().background(Brand.pink).padding(horizontal = 14.dp, vertical = 12.dp),
                verticalAlignment = Alignment.CenterVertically,
            ) {
                Text("✕", fontSize = 17.sp, fontWeight = FontWeight.Bold, color = Color.White,
                    modifier = Modifier.clickable { onDismiss() }.padding(6.dp))
                Spacer(Modifier.width(10.dp))
                Text("Schedules", fontSize = 18.sp, fontWeight = FontWeight.Bold, color = Color.White,
                    modifier = Modifier.weight(1f))
                if (saving) Text("Saving…  ", fontSize = 11.sp, color = Color.White.copy(alpha = 0.8f))
                Text("＋ Add", fontSize = 15.sp, fontWeight = FontWeight.Bold, color = Color.White,
                    modifier = Modifier.clickable { showNew = true }.padding(6.dp))
            }

            val r = rows
            Column(Modifier.fillMaxSize().verticalScroll(rememberScrollState()).padding(16.dp)) {
                when {
                    r == null -> LoadingSpinner("Loading…")
                    r.isEmpty() -> Text(
                        "No scheduled prompts yet. Add a reminder with ＋, or build richer prompts (questions, game nudges) on the web dashboard.",
                        fontSize = 13.sp, color = Brand.muted,
                    )
                    else -> r.forEachIndexed { idx, row ->
                        ScheduleRow(
                            row = row,
                            onToggle = { enabled ->
                                val updated = r.toMutableList()
                                updated[idx] = buildJsonObject {
                                    for ((k, v) in row) if (k != "enabled") put(k, v)
                                    put("enabled", enabled)
                                }
                                save(updated)
                            },
                            onDelete = { save(r.filterIndexed { i, _ -> i != idx }) },
                        )
                        Spacer(Modifier.height(10.dp))
                    }
                }
            }
        }
    }

    if (showNew) {
        NewReminderDialog(
            onAdd = { prompt, intervalMin ->
                showNew = false
                val row = buildJsonObject {
                    put("id", "s${System.currentTimeMillis()}")
                    put("type", "reminder")
                    put("enabled", true)
                    put("prompt", prompt)
                    put("timing", "interval")
                    put("intervalMin", intervalMin)
                    put("days", buildJsonArray { (0..6).forEach { add(JsonPrimitive(it)) } })
                }
                save((rows ?: emptyList()) + row)
            },
            onDismiss = { showNew = false },
        )
    }
}

@Composable
private fun ScheduleRow(row: JsonObject, onToggle: (Boolean) -> Unit, onDelete: () -> Unit) {
    val type = (row["type"] as? JsonPrimitive)?.content ?: "reminder"
    val enabled = try { (row["enabled"] as? JsonPrimitive)?.boolean ?: true } catch (_: Exception) { true }
    val prompt = (row["prompt"] as? JsonPrimitive)?.content?.takeIf { it.isNotEmpty() }
        ?: when (type) {
            "question" -> "A quick question."
            "game" -> "Let's do a game!"
            else -> "Time for a check-in."
        }
    val typeLabel = when (type) { "question" -> "Question"; "game" -> "Game nudge"; else -> "Reminder" }
    val emoji = when (type) { "question" -> "❓"; "game" -> "🎮"; else -> "🔔" }
    val timing = run {
        val timingKind = (row["timing"] as? JsonPrimitive)?.content
        val times = (row["times"] as? JsonArray)?.mapNotNull { (it as? JsonPrimitive)?.content }
        if (timingKind == "times" && !times.isNullOrEmpty()) "at " + times.joinToString(", ")
        else {
            val mins = (row["intervalMin"] as? JsonPrimitive)?.doubleOrNull ?: 45.0
            "every ${mins.toInt()} min"
        }
    }

    Row(
        Modifier.fillMaxWidth()
            .background(Color.White, RoundedCornerShape(14.dp))
            .border(1.dp, hexColor("#f3c6da"), RoundedCornerShape(14.dp))
            .padding(12.dp),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        Text(emoji, fontSize = 20.sp)
        Spacer(Modifier.width(10.dp))
        Column(Modifier.weight(1f)) {
            Text(prompt, fontSize = 15.sp, fontWeight = FontWeight.SemiBold,
                color = if (enabled) Brand.ink else Brand.muted)
            Text("$typeLabel · $timing", fontSize = 12.sp, color = Brand.muted)
        }
        TextButton(onClick = onDelete) { Text("🗑", fontSize = 15.sp) }
        Switch(checked = enabled, onCheckedChange = onToggle,
            colors = SwitchDefaults.colors(checkedTrackColor = Brand.pink))
    }
}

/** Minimal composer for a spoken interval reminder ("Do you need the potty?"). */
@Composable
private fun NewReminderDialog(onAdd: (String, Int) -> Unit, onDismiss: () -> Unit) {
    var prompt by remember { mutableStateOf("") }
    var intervalMin by remember { mutableStateOf(45) }

    AlertDialog(
        onDismissRequest = onDismiss,
        title = { Text("New reminder") },
        text = {
            Column {
                OutlinedTextField(
                    value = prompt, onValueChange = { prompt = it },
                    label = { Text("What should the board say?") },
                    modifier = Modifier.fillMaxWidth(),
                )
                Spacer(Modifier.height(10.dp))
                Row(verticalAlignment = Alignment.CenterVertically) {
                    TextButton(onClick = { if (intervalMin > 5) intervalMin -= 5 }) { Text("−", fontSize = 20.sp) }
                    Text("Every $intervalMin minutes", fontSize = 14.sp, modifier = Modifier.weight(1f),
                        fontWeight = FontWeight.SemiBold)
                    TextButton(onClick = { if (intervalMin < 240) intervalMin += 5 }) { Text("＋", fontSize = 18.sp) }
                }
            }
        },
        confirmButton = {
            TextButton(
                onClick = { onAdd(prompt.trim(), intervalMin) },
                enabled = prompt.isNotBlank(),
            ) { Text("Add", fontWeight = FontWeight.Bold) }
        },
        dismissButton = { TextButton(onClick = onDismiss) { Text("Cancel") } },
    )
}
